From 06636162b1c869281f17bf00011659dee1a6366c Mon Sep 17 00:00:00 2001
From: Yotam Ben Dosa <yotambd@linux4biz.com>
Date: Mon, 5 Sep 2016 17:51:24 +0300
Subject: [PATCH] srvkm:module: Check whether Soc supports SGX

Patch is used to ensure that devices that can't use sgx
won't be able to load the pvrsrvkm module.

Original author:
Teresa Remmet <t.remmet@phytec.de>

Signed-off-by: Yotam Ben Dosa <yotambd@linux4biz.com>
---
 eurasia_km/services4/srvkm/env/linux/module.c | 30 ++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/eurasia_km/services4/srvkm/env/linux/module.c b/eurasia_km/services4/srvkm/env/linux/module.c
index fd61752..0044f40 100644
--- a/eurasia_km/services4/srvkm/env/linux/module.c
+++ b/eurasia_km/services4/srvkm/env/linux/module.c
@@ -973,7 +973,35 @@ static int __init PVRCore_Init(void)
	struct device *psDev;
 #endif

-
+        u32 r;
+        void __iomem *omap2_ctrl_base;
+        const u32 iobase_ctrl_module_feat = 0x44E10604;
+        const u32 sgx_shift = 29;
+
+        if (!request_mem_region(iobase_ctrl_module_feat, 0x4, "ctrl-test")) {
+                printk(KERN_INFO "request_mem_region failed");
+                error = -EFAULT;
+                return error;
+        }
+
+        omap2_ctrl_base = ioremap(iobase_ctrl_module_feat, 0x4);
+        if (!omap2_ctrl_base) {
+                printk(KERN_INFO "ioremap failed");
+                release_mem_region(iobase_ctrl_module_feat, 0x4);
+                error = -EFAULT;
+                return error;
+        }
+
+        r = __raw_readl(omap2_ctrl_base);
+
+        iounmap(omap2_ctrl_base);
+        release_mem_region(iobase_ctrl_module_feat, 0x4);
+
+        if ((r & (1 << sgx_shift)) == 0) {
+                printk(KERN_ERR "No SGX on SoC found.\n");
+                error = -ENODEV;
+                return error;
+        }

 #if !defined(SUPPORT_DRI_DRM)
	/*
--
1.9.1
